README.md

  # Login Widget  
 Show content depending on whether or not the user is logged in. Simply provide two arguments: A Widget to display when logged in, one to display when not.  The goal is to make setting up a login page and handling that process *much* less of a hassle, and I am shaming Google hardcore for not building *something* like this this for the core framework or the FlutterFire package.
  
*Relies on the firebase_auth and google_sign_in packages. This was written in about two hours to save time going forward, so pull requests are welcome if you can add to the package.*
  
## Usage  
  
The Login widget is meant to simplify the process of stateful content for logged-in / logged-out users, and it was designed to be as easy to use as possible. There are a few static methods (like `Login.signInWithGoogle` and `Login.signInWithEmail`)  that work out of the box, but you can use any of the login methods provided by `FirebaseAuth.instance` and easily verify there were no issues with `Login.checkUser(FirebaseUser user)`.  
  
Declaring a Login widget:  
```  
Login(
    loggedIn: myHomePage(), 
	loggedOut: myLoginForm(), 
)
```
Yup. That's it.  

## Sign in with Google
Inside `myLoginForm` (or any widget), you could simply call `Login.signInWithGoogle()` to call the google_sign_in plugin and automatically refresh the page when the user state changes.  This is especially helpful for buttons, i.e., inside `myLoginForm`:  
```  
RaisedButton(onPressed: Login.signInWithGoogle, child: Text("Sign in"))  
```  
For this specific case, the work has been done for you, and the package includes a helper widget with the onPressed set to the static method `Login.signInWithGoogle`. 
```
GoogleSignInButton()
```

Effortless. You can initialize this with *no arguments*, as seen above. (Take notes, Mr. Sneath.)   All of the branding guidelines provided by Google have been incorporated into the widget.

![enter image description here](https://developers.google.com/identity/images/google_button_spec.png)

The `Login` Widget will automatically display the relevant content when the login status changes using the `FirebaseAuth.instance.onAuthStateChanged` listener. In the first example under the *Usage* section, `myHomePage` will be shown when the user is logged in, and `myLoginForm` will be displayed when the user is not logged in.
  
Easy peeze. Additionally, because each app will (generally) only have one user, there is no harm (and tons of convenience) in exposing the current Firebase user data and login status as static properties `(FirebaseUser) Login.currentUser` and `(bool) Login.isLoggedIn`. You can access these at any time anywhere in your code.  

## Login Fields
There are now standard static TextFields available at `Login.email` and `Login.password`. The controller logic is handled by the Login package. Reference these static widgets wherever you want to accept an email or password for logging in, and customize the styling in your app's `ThemeData`.

`Login.signInWithEmail` will automatically read the text from its private TextEditingControllers (email and password) for the sign-in process, same with `Login.resetPassword`.  These static TextFields will share the same controllers no matter where they're initialized, which can be an inconvenience but can also be useful - for instance, initializing *both* a sign-in form and a "forgot password" form with `Login.email` will mean text is shared across the two fields. A user who types their email to sign in, and then opens the Forgot Password page, will automatically have that text inputted for them - they are, quite literally, the same field. 

Before, these functions accepted TextEditingControllers when invoked, but this package is really meant to be as simple to use as possible so those parameters have been removed.  Simply place these static widgets somewhere and call `Login.signInWithEmail()` on a button press to attempt the sign-in.

## Automatic Snackbars
Most of the sign-in functions accept a `BuildContext context` to trigger a relevant Snackbar message. For instance, calling `Login.resetPassword(context: context)` will show a message saying "Check your email for a password reset link." Failed logins will behave similarly.

## Theming
Any theming you wish to do can be done in your app's `ThemeData`, including the email and password TextFields and Snackbars.
  
*Pre-Installation & Debugging*  
  
**1. Make sure you set up Firebase before using the widget, otherwise FirebaseAuth will not work.** For instance, the example included in this package displays the usage and general concept, but cannot actually change state without a google-services file or the [necessary configuration](https://firebase.google.com/docs/flutter/setup). Once Firebase is set up, you can simply depend on the Login package and it will work automatically.  
  
**2. Make sure your Firebase project is [set up](https://console.firebase.google.com) with a support email in Project Settings or you may get backend errors.**  
  
**3. Make sure you have sign-in methods enabled under Authentication.**

CHANGELOG.md
## [1.0.0] - 10/14/2019

Initial release.

## [1.0.1] - 10/15/2019

Added BuildContext params, for easy error handling.

## [1.1.0] - 10/18/2019

Added static Login.email and Login.password TextFields, plus Snackbars.

## [1.1.1] - 10/18/2019

Double-checked Google Sign In branding guidelines.

test/login_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:login/login.dart';

void main() {
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:login_example/main.dart';

void main() {
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:login/login.dart';

void main() => runApp(myApp());

class myApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Login Example",
      home: Scaffold(
          body: Login(
            loggedIn: myHomePage(),
            loggedOut: myLoginForm(),
          )
      ),
    );
  }
}

class myHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Text("You're logged in!"),
    );
  }
}

class myLoginForm extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisSize: MainAxisSize.max,
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: Text("You're not logged in! \n\n Sign in won't actually work on this page because Firebase is not set up, but it does illustrate how the package works out of the box. Tap the button to try it out! \n\n If FirebaseAuth received a login update containing a non-null FirebaseUser, this page would automatically refresh to show myHomePage().", textAlign: TextAlign.center,),
          ),
          GoogleSignInButton()
        ]
      )
    );
  }
}


lib/login.dart
library login;

import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:google_sign_in/google_sign_in.dart';

class Login extends StatefulWidget {
  Login({@required this.loggedIn, @required this.loggedOut});

  final Widget loggedIn;
  final Widget loggedOut;

  static var currentUser;
  static var isLoggedIn = false;

  static final TextEditingController _emailController = TextEditingController();
  static final TextEditingController _passwordController = TextEditingController();

  static final Widget email = TextField(
      controller: _emailController,
      keyboardType: TextInputType.emailAddress,
      decoration: InputDecoration(labelText: "Email")
  );

  static final Widget password = TextField(
      controller: _passwordController,
      obscureText: true,
      decoration: InputDecoration(labelText: "Password")
  );

  static Future<bool> checkLogin(FirebaseUser user) async {
    try {
      assert(user.email != null);
      assert(user.displayName != null);
      assert(!user.isAnonymous);
      assert(await user.getIdToken() != null);

      final FirebaseUser currentUser =
          await FirebaseAuth.instance.currentUser();
      assert(user.uid == currentUser.uid);
    } catch (e) {
      debugPrint("Failure logging in.");
      return false;
    }

    return true;
  }

  static Future<FirebaseUser> signUpWithEmail(
      {@required TextEditingController email,
      @required TextEditingController password,
      BuildContext context}) async {
    final String _email = email.text.trim();
    final String _password = password.text;

    AuthResult authResult =
        await FirebaseAuth.instance.createUserWithEmailAndPassword(
      email: _email,
      password: _password,
    );

    try {
      final FirebaseUser _user = authResult.user;
      await _user.sendEmailVerification();
      return _user;
    } catch (e) {
      debugPrint("Error creating user.");
      return null;
    }
  }

  static Future<void> resetPassword({BuildContext context}) async {
    if (_emailController.text.isEmpty) return null;
    Navigator.of(context).pop();

    if (context != null)
      showSnackbar(
        msg: "Check your email for a password reset link.",
        context: context,
      );

    await FirebaseAuth.instance
        .sendPasswordResetEmail(email: _emailController.text.trim());
  }

  static Future<FirebaseUser> signInWithEmail({BuildContext context}) async {
    try {
      final String _email = _emailController.text.trim();
      final String _password = _passwordController.text;
      final AuthResult result = await FirebaseAuth.instance
          .signInWithEmailAndPassword(email: _email, password: _password);
      final FirebaseUser user = result.user;

      await checkLogin(user);

      return user;
    } catch (e) {
      debugPrint("Error: $e");

      if (context == null) return null;
      error(email: _emailController, password: _passwordController, context: context);
      return null;
    }
  }

  static Future<FirebaseUser> signInWithGoogle({BuildContext context}) async {
    try {
      final GoogleSignInAccount googleUser = await GoogleSignIn().signIn();
      final GoogleSignInAuthentication googleAuth =
          await googleUser.authentication;
      final AuthCredential credential = GoogleAuthProvider.getCredential(
        accessToken: googleAuth.accessToken,
        idToken: googleAuth.idToken,
      );
      final FirebaseUser user =
          (await FirebaseAuth.instance.signInWithCredential(credential)).user;

      await checkLogin(user);

      return user;
    } catch (e) {
      debugPrint("Error: $e");
      return error(context: context);
    }
  }

  static signOut() {
    FirebaseAuth.instance.signOut();
  }

  static Future error(
      {String msg = "Login failed. Please try again.",
      TextEditingController email,
      TextEditingController password,
      @required BuildContext context}) {
    if (email != null) email.clear();
    if (password != null) password.clear();

    return showSnackbar(msg: msg, context: context);
  }

  static Future showSnackbar(
      {@required BuildContext context, String msg, SnackBar snackBar}) {
    Scaffold.of(context).showSnackBar(
        snackBar == null ? SnackBar(content: Text(msg)) : snackBar);
    return null;
  }

  @override
  _LoginState createState() => _LoginState();
}

class _LoginState extends State<Login> {
  final FirebaseAuth _instance = FirebaseAuth.instance;

  @override
  void initState() {
    super.initState();
    try {
      _instance.onAuthStateChanged.listen((FirebaseUser firebaseUser) {
        setState(() {
          Login.currentUser = firebaseUser;
          Login.isLoggedIn = firebaseUser != null;
        });
      });
    } catch (e) {
      debugPrint("Error: $e");
    }
  }

  @override
  Widget build(BuildContext context) {
    return Login.isLoggedIn ? widget.loggedIn : widget.loggedOut;
  }
}

class GoogleSignInButton extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      height: 40.0,
      child: RaisedButton(
        elevation: 8,
        splashColor: Colors.grey,
        color: Colors.white,
        onPressed: Login.signInWithGoogle,
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(40)),
        child: Padding(
          padding: const EdgeInsets.fromLTRB(0, 8, 0, 8),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Image(
                  image: AssetImage("assets/google.png", package: "login"),
                  height: 18),
              Container(
                width: 24.0,
              ),
              Text(
                'SIGN IN WITH GOOGLE',
                style: TextStyle(
                  fontSize: 14,
                  fontFamily: "Roboto",
                  fontWeight: FontWeight.w500,
                  color: Colors.black54,
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
